design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,483 @@
+  name: frontend-a75588698
+  namespace: default
+  ownerReferences:
+  - apiVersion: core.oam.dev/v1alpha2
+    blockOwnerDeletion: true
+    controller: true
+    kind: Component
+    name: frontend
+revision: 2
+data:
+  spec:
+    workload:
+      apiVersion: core.oam.dev/v1alpha2
+      kind: ContainerizedWorkload
+      spec:
+        containers:
+        - name: my-cool-workload
+          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+          cmd:
+          - "bash top"
+```
+
+Now the basic controller revision are in place, let's talk about them in Trait and ApplicationConfiguration.
+
+### ControllerRevision could be referenced by Trait 
+
+Some traits can control version of component which means they can specify a component revision and won't be affected
+by the change of component.
+
+For these traits, they have to declare `revisionAware` field to be `true` in TraitDefinition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: fancytraits.core.oam.dev
+spec:
+  revisionAware: true
+  definitionRef:
+    name: fancytraits.core.oam.dev
+```
+
+Versioning mechanism of OAM runtime will be according to the  `revisionAware` field.
+
+- If **NO** traits binding with a component has `revisionAware` as true, OAM runtime will let the workload itself to handle
+   versioning things. We will only have one workload instance with the same name of componentName, and update workload directly
+   when component changed.
+- If **ANY** trait binding with a component has `revisionAware` as true, OAM runtime will always create a new workload with
+   a new name which can be aligned with the name of newly created component revision. OAM runtime will leave the trait to
+   control existing workloads to do rolling update and garbage collection. 
+
+So when we don't use any trait who has revisionAware as true, things will work like below:
+
+If we create an `ApplicationConfiguration` with definition : 
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+```
+
+The real workload instance will be created with the same name as the componentName.
+
+```yaml
+  apiVersion: core.oam.dev/v1alpha2
+  kind: ContainerizedWorkload
+  metadata:
+    name: frontend
+  spec:
+    containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+      - "bash lscpu"
+```
+
+But if we use any trait has revisionAware like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: FancyTrait
+            spec:
+              traffic:
+                - revision: frontend-c8bb659c5
```

Comment:
Agreed with opinions from @hasheddan , I would not prefer to specify a component `reversion` value for an `ApplicationConfiguration`. Especially in a situation when an `ApplicationConfiguration` already exists, I need to modify it or do deletion and recreation operation.

Reply:
see https://github.com/crossplane/oam-kubernetes-runtime/pull/29#discussion_r432265873